core/src/main/scala/kafka/server/DelayedFetch.scala

```diff
@@ -114,6 +123,19 @@ class DelayedFetch(delayMs: Long,
                   accumulatedSize += bytesAvailable
               }
             }
+
+            // Case G check HW from follower against latest from leader
+            val followerHW = followerHighwatermarks.apply(topicPartition)
```

Comment:
In the KIP we proposed that we would return immediately from the first fetch in a session in order to be sure that the follower had an updated high watermark.

also nit: no need to call `apply` directly.

Reply:
Ok, so if we have a request from a follower and no HW information yet (meaning it's the first request in the session), we should just return immediately?

Reply:
Yeah, does it make sense? Since we're tracking the high watermark in the fetch session, we lose the latest value if we lose the session. So when the session is initialized, we assume the follower doesn't have the latest value. We could probably get around this by tracking the high watermark in the `Replica` object instead. Not sure how easy that would be.

I guess this case should probably be handled in the first fetch in `ReplicaManager.fetchMessages`. So it never actually hits `DelayedFetch`.

Reply:
Oh, there's one thing we probably have to be careful about. If we are upgrading from an older version, the other broker may not have a fetch session. It would probably be reasonable to skip the high watermark check in this case. Unfortunately, this might mean passing down another flag or something.